### PR TITLE
Refactor image features selection in LlaVa

### LINKED ISSUE #33695: Refactor image features selection in LlaVa
### Feature request

We would like to wrap image features selection in LLaVa in a separate function to make it easier to override for custom use cases (e.g. applying a layer norm on the image features before projection, etc).

https://github.com/huggingface/transformers/blob/52daf4ec768fb9ffe84a0c373834172a7c54aecc/src/transformers/models/llava/modeling_llava.py#L452-L461

Currently, we need to override the entire forward pass for any custom processing of the image features. Having a separate function would make it a lot easier! 

### Motivation

To make it easier to implement custom image features selection in LlaVa. 

### Your contribution

A custom function for image features selection
```python
    def _get_selected_image_features(
        self, pixel_values: torch.FloatTensor, vision_feature_layer: int, vision_feature_select_strategy: str
    ):
        image_outputs = self.vision_tower(pixel_values, output_hidden_states=True)
        # this is not memory efficient at all (output_hidden_states=True) will save all the hidden stated.
        selected_image_feature = image_outputs.hidden_states[vision_feature_layer]
        if vision_feature_select_strategy == "default":
            selected_image_feature = selected_image_feature[:, 1:]
        elif vision_feature_select_strategy == "full":
            selected_image_feature = selected_image_feature
        else:
            raise ValueError(f"Unexpected select feature strategy: {self.config.vision_feature_select_strategy}")
        return selected_image_feature
```

### PR DESCRIPTION
**Pull request by kenza-bouzid**
# What does this PR do?

Wrap image features selection in LLaVa in a separate function to make it easier to override for custom use cases (e.g. applying a layer norm on the image features before projection, etc).

Fixes #33695 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@ArthurZucker, @younesbelkada

### CODE REVIEW

**Review by zucchini-nlp — COMMENTED** [review]
Hey @kenza-bouzid !

Thanks for opening a PR! Yes, actually it would be nice to have as a rule that VLMs have a method to `get_image_features`, where all logic happens. My prev idea was to make code less cluttered when the VLM has many special crops/pads/concats. But the use-case of overriding the method for custom models is very coool

We already have a similar pattern in video LLMs, for the same reason of uncluttering the code. WDYT about moving the whole image feature preparation to a separate method, including MM-proj. So the method takes in pixels and return features ready to be merged with text embeddings?

And, also, we might need to change more models to follow the same pattern to make CI happy. If you have bandwidth, would be nice to propagate change to all VLMs. But don't stress if you can't, I'll make an overall standardization soon on that :)

**Comment by kenza-bouzid** [comment]
Hi @zucchini-nlp, 

Thanks for your reply. 

>WDYT about moving the whole image feature preparation to a separate method, including MM-proj. So the method takes in pixels and return features ready to be merged with text embeddings?

Good idea, however I would suggest having a separate method for that as well to make it easy to customize the projection. I can think of many use cases where you want to easily change the way you project.
  
We can have `get_image_features` and `project_image_features`  that are then called sequentially in a `get_image_embeddings` ready to be merged with text embeddings as you suggested. 

> If you have bandwidth, would be nice to propagate change to all VLMs. But don't stress if you can't, I'll make an overall standardization soon on that :)

I'm afraid I won't have time to propagate the change to all VLLMs. Is that a blocker for this PR to pass the CI? You probably have more context for an overall standardization!

**Comment by zucchini-nlp** [comment]
> I would suggest having a separate method for that as well to make it easy to customize the projection. I can think of many use cases where you may want to change the projection.

Hmm, imo that is a bit redundant, and since "select+proj" is mostly interlinked and isn't a huge piece of code we may ask users to override that one method, Independently of whether they tweak at selection stage or projection stage. So that we don't end up with separate methods that perform a one-liner, which imo is a bad practice. Lmk if you have any other ideas or objections

> I'm afraid I won't have time to propagate the change to all VLLMs. Is that a blocker for this PR to pass the CI? You probably have more context for an overall standardization!

No worries, I just realized it is the very first llava being modified here. So our CI should be happy after running `make fix-copies` and `make style`. For other VLMs that are more tricky, I'll make one round of standardization later. Not a blocker at all  :)

**Comment by kenza-bouzid** [comment]
Sounds great! Let me make the changes you suggested. Thank youu!

File: src/transformers/models/vipllava/modeling_vipllava.py
Comment (kenza-bouzid) [inline src/transformers/models/vipllava/modeling_vipllava.py:285]: @zucchini-nlp `make fix-copies` copied this over but looking at the forward pass of vipllava, the image features selection is slightly different since they select features from for the layers -2, -5, -8, -11 and 6
I don't think it's breaking anything since this function is not called. Note that we'll need ` vision_feature_layers: list[int]` and not only a single one!

Comment (zucchini-nlp) [reply src/transformers/models/vipllava/modeling_vipllava.py:285]: Hmm, in that case better to make the method accept list in vipllava and add a "#Ignore copy" statement so that it is not copied from llava. That way we can use the `get_image_features` directly

I don't think it's a good idea to merge smth which doesn't work, even tho it technically doesn't break anything

**Review by zucchini-nlp — APPROVED** [review]
Thanks a lot! LGTM in general, left one comment for vipllava

I'm approving this PR, when you're ready with vipllava feel free to tag @ LysandreJik (core maintainer). After his approval, PR can be merged :)

File: src/transformers/models/vipllava/modeling_vipllava.py
Comment (kenza-bouzid) [reply src/transformers/models/vipllava/modeling_vipllava.py]: agreed 💯 fixed now in https://github.com/huggingface/transformers/pull/33696/commits/f75e94470fd8d88b0e8e702826728660a9e618a9

**Comment by kenza-bouzid** [comment]
@LysandreJik can you please have a look? Thanks

**Review by ArthurZucker — APPROVED** [review]
LGTM, the forward pass is pretty long, and this will come handy with modular soon!

**Pull request merged**
